File: 05-FEA-Heat/Matlab-Code/EvaluateFieldGradient.m
function [FluxAtIntegrationPoints, ElementFlux, NodalFlux, MagnitudeNodalFlux,...
            NodalGradT, MagnitudeGradT] = ...
            EvaluateFieldGradient(N_ELEM, N_NODE, NNODE_ELE, DOF_NODE, ...
            NDIM, COORDS, ELEM_NODE, ELEM_STIFF, ShapeOrder , Ng, UUR)

%% Get the Flux at the integration points
location = 1; % evaluate at the gauss points used for the solution
FluxAtIntegrationPoints = zeros(N_ELEM*NDIM*Ng,NDIM);

for I_ELEM=1:N_ELEM
    [flux, gradT] = GetElementHeatFluxVector(I_ELEM, NNODE_ELE, DOF_NODE, ...
            NDIM, COORDS, ELEM_NODE, ELEM_STIFF, ShapeOrder , Ng, location, UUR);
        for j=1:NDIM*Ng
        pos =  (I_ELEM-1)*NDIM*Ng + j;
        FluxAtIntegrationPoints(pos,:)=flux(:,j);
     end
end

%% Get the Flux at the centroid for extrapolation to nodes
location = 0; % evaluate at the center of the element
ElementFlux = zeros(N_ELEM,NDIM);

for I_ELEM=1:N_ELEM
    [flux, gradT] = GetElementHeatFluxVector(I_ELEM, NNODE_ELE, DOF_NODE, ...
            NDIM, COORDS, ELEM_NODE, ELEM_STIFF, ShapeOrder , Ng, location, UUR);
    ElementFlux(I_ELEM,:)=flux'; 
    ElementGradT(I_ELEM,:)=gradT';
end

%% Extrapolate flux at nodes

scale = zeros(N_NODE,1); % vector containing the number of elements attached to each node
NodalFlux = zeros(N_NODE,NDIM); % matrix with the nodal values of the heat flux vector
NodalGradT = zeros(N_NODE,NDIM);
MagnitudeNodalFlux = zeros(N_NODE,1);
MagnitudeGradT = zeros(N_NODE,1);

for I_ELEM=1:N_ELEM
    for j=1:NNODE_ELE
        node = ELEM_NODE(j,I_ELEM);
        scale(node) = scale(node) + 1;
        NodalFlux(node,:) = NodalFlux(node,:) + ElementFlux(I_ELEM,:);
        NodalGradT(node,:)  = NodalGradT(node,:)  + ElementGradT(I_ELEM,:);
    end
end


for i=1:N_NODE
    for j=1:NDIM
        NodalFlux(i,j) = NodalFlux(i,j)/scale(i);
        NodalGradT(i,j) = NodalGradT(i,j)/scale(i);
    end
    MagnitudeNodalFlux(i) = sqrt(NodalFlux(i,1)^2+NodalFlux(i,2)^2);
    MagnitudeGradT(i) = sqrt(NodalGradT(i,1)^2+NodalGradT(i,2)^2);
end

File: 05-FEA-Heat/C-Code/EvaluateFieldGradient.cpp
#include <MainFile.h>
using namespace std;


void EvaluateFieldGradient( const int nnode_ele, const int n_node, const int n_elem, const int dof_node,
		const int Ng, const int ShapeOrder, const int ndim, const arma::mat& coords, const arma::imat& elem_node,
		const arma::mat& elem_stiff, const arma::mat& Uur,
		arma::mat& FluxAtIntegrationPoints, arma::mat& ElementFlux, arma::mat& NodalFlux,
		arma::colvec& MagnitudeNodalFlux, arma::mat& NodalGradT, arma::colvec& MagnitudeGradT)
		
{

	  FluxAtIntegrationPoints = arma::zeros<arma::mat>(n_elem*ndim*Ng,ndim);
	  ElementFlux = arma::zeros<arma::mat>(n_elem,ndim);
	  NodalFlux = arma::zeros<arma::mat>(n_node,ndim);
	  MagnitudeNodalFlux = arma::zeros<arma::colvec>(n_node);
	  NodalGradT = arma::zeros<arma::mat>(n_node,ndim);
	  MagnitudeGradT = arma::zeros<arma::colvec>(n_node);

	  // ------------------------------------------------------------
	  // ------- Get the Flux at the integration points -------------
	  // ------------------------------------------------------------
	  int location = 1; // evaluate at the gauss points used for the solution
	  arma::mat flux = arma::zeros<arma::mat>(ndim,Ng*ndim);
	  arma::mat gradT = arma::zeros<arma::mat>(ndim,Ng*ndim);

	  for (int i=0;i<n_elem;i++) {
	  	int iel = i + 1;
		GetElementHeatFluxVector(iel,nnode_ele, dof_node, Ng, ShapeOrder, ndim, coords, elem_node,
					elem_stiff, location , Uur, flux, gradT);
		for (int j=0;j<ndim*Ng;j++){
			int pos = (i)*ndim*Ng + j;
			FluxAtIntegrationPoints.row(pos) = arma::trans(flux.col(j));
		  }
	  }
		cout << "Finish flux at integration points" << endl;
	  // ------------------------------------------------------------
	  // ------- Get the Flux at the centroid -----------------------
	  // ------------------------------------------------------------
	  location = 0; // evaluate at the gauss points used for the solution
	  arma::mat fluxC = arma::zeros<arma::mat>(ndim,1);
	  arma::mat gradTC = arma::zeros<arma::mat>(ndim,1);
	  arma::mat ElementGradT = arma::zeros<arma::mat>(n_elem,ndim);

	  for (int i=0;i<n_elem;i++) {
	  	  int iel = i + 1;
		  GetElementHeatFluxVector(iel,nnode_ele, dof_node, Ng, ShapeOrder, ndim, coords, elem_node,
					elem_stiff, location , Uur, fluxC, gradTC);
		  ElementFlux.row(i) = arma::trans(fluxC);
		  ElementGradT.row(i) = arma::trans(gradTC);
	  }

	  // ------------------------------------------------------------
	  // ------- Extrapolate flux to nodes    -----------------------
	  // ------------------------------------------------------------
	  arma::colvec scale = arma::zeros<arma::mat>(n_node);
	  int node;
	  for (int i=0;i<n_elem;i++) {
		  for (int j=0;j<nnode_ele;j++) {
			  node = elem_node(j,i)-1;
			  scale(node) = scale(node) + 1;
			  NodalFlux.row(node) = NodalFlux.row(node) + ElementFlux.row(i);
			  NodalGradT.row(node) = NodalGradT.row(node) + ElementGradT.row(i);
		  }
	  }
	  for (int i=0;i<n_node;i++) {
	  		  for (int j=0;j<ndim;j++) {
	  			  NodalFlux(i,j) = NodalFlux(i,j)/scale(i);
	  			  NodalGradT(i,j) = NodalGradT(i,j)/scale(i);
	  		  }
	  		  MagnitudeNodalFlux(i) = arma::norm( NodalFlux.row(i) , 2 );
	  		  MagnitudeGradT(i) = arma::norm( NodalGradT.row(i) , 2 );
	  }

  
}
